File: chp2/python/WeatherData.py
from Subject import Subject


class WeatherData(Subject):
    def __init__(self):
        super(WeatherData, self).__init__()
        self.observers = []

    def registerObserver(self, observer):
        self.observers.append(observer)

    def removeObserver(self, observer):
        self.observers.remove(observer)

    def notifyObserver(self):
        for observer in self.observers:
            observer.update(self.temperature, self.humidity, self.pressure)

    def measurementsChanged(self):
        self.notifyObserver()

    def setMeasurements(self, temperature, humidity, pressure):
        self.temperature = temperature
        self.humidity = humidity
        self.pressure = pressure
        self.measurementsChanged()

File: chp2/java/src/WeatherData.java
import java.util.*;


public class WeatherData implements Subject {
    private float tempterature;
    private float humidity;
    private float pressure;
    private ArrayList observers;

    public WeatherData() {
	observers = new ArrayList();
    }

    public void registerObserver(Observer o) {
	observers.add(o);
    }

    public void removeObserver(Observer o) {
	int i = observers.indexOf(o);
	if (i >= 0) {
	    observers.remove(i);
	}
    }

    public void notifyObserver() {
	for (int i = 0; i < observers.size(); i++) {
	    Observer observer = (Observer)observers.get(i);
	    observer.update(tempterature, humidity, pressure);
	}
    }

    public void measurementsChanged() {
	notifyObserver();
    }

    public void setMeasurements(float tempterature, float humidity, float pressure) {
	this.tempterature = tempterature;
	this.humidity = humidity;
	this.pressure = pressure;
	measurementsChanged();
    }
}
